File: chroma-league/src/test/java/com/bonepl/chromaleague/IntSteps.java
package com.bonepl.chromaleague;

public class IntSteps {
    private final int from;
    private final int to;
    private final int step;
    private int current;
    private int direction = 1;

    public IntSteps(int from, int to, int step) {
        this.from = from;
        this.to = to;
        this.step = step;
        current = from;
    }

    public int nextInt() {
        if (current + step > to || current + step < from) {
            direction = Math.negateExact(direction);
        }
        current += step * direction;
        return current;
    }
}

File: test/IntSteps.ts
export class IntSteps {
    private readonly from: number;
    private readonly to: number;
    private readonly step: number;
    private current: number;
    private direction = 1;

    constructor(from: number, to: number, step: number) {
        this.from = from;
        this.to = to;
        this.step = step;
        this.current = from;
    }

    nextInt(): number {
        const toReturn = this.current;
        const nextIncrement = this.step * this.direction;
        if (this.from < this.to) {
            if (this.current + nextIncrement > this.to || this.current + nextIncrement < this.from) {
                this.direction = -this.direction;
            }
        } else {
            if (this.current + nextIncrement < this.to || this.current + nextIncrement > this.from) {
                this.direction = -this.direction;
            }
        }
        this.current += this.step * this.direction;
        return toReturn;
    }
}
